src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -439,70 +442,80 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
         internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
         {
-            if (CommonUtil.IsUnsupportedType(type))
-            {
-                throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
-            }
+            ODataTypeInfo typeInfo = GetODataTypeInfo(type);
 
-            string typeName = type.ToString();
-            IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
-            bool isEntity = customAttributes.OfType<EntityTypeAttribute>().Any();
-            KeyAttribute dataServiceKeyAttribute = customAttributes.OfType<KeyAttribute>().FirstOrDefault();
-            List<PropertyInfo> keyProperties = new List<PropertyInfo>();
-            PropertyInfo[] properties = ClientTypeUtil.GetPropertiesOnType(type, false /*declaredOnly*/).ToArray();
- 
-            hasProperties = properties.Length > 0;
-            KeyKind currentKeyKind = KeyKind.NotKey;
-            KeyKind newKeyKind = KeyKind.NotKey;
-            foreach (PropertyInfo propertyInfo in properties)
+            if (typeInfo.KeyProperties == null)
             {
-                if ((newKeyKind = ClientTypeUtil.IsKeyProperty(propertyInfo, dataServiceKeyAttribute)) != KeyKind.NotKey)
+                if (CommonUtil.IsUnsupportedType(type))
                 {
-                    if (newKeyKind > currentKeyKind)
-                    {
-                        keyProperties.Clear();
-                        currentKeyKind = newKeyKind;
-                        keyProperties.Add(propertyInfo);
-                    }
-                    else if (newKeyKind == currentKeyKind)
-                    {
-                        keyProperties.Add(propertyInfo);
-                    }
+                    throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
                 }
-            }
 
-            Type keyPropertyDeclaringType = null;
-            foreach (PropertyInfo key in keyProperties)
-            {
-                if (keyPropertyDeclaringType == null)
-                {
-                    keyPropertyDeclaringType = key.DeclaringType;
-                }
-                else if (keyPropertyDeclaringType != key.DeclaringType)
+                string typeName = type.ToString();
+                IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
+                bool isEntity = customAttributes.OfType<EntityTypeAttribute>().Any();
+                KeyAttribute dataServiceKeyAttribute = customAttributes.OfType<KeyAttribute>().FirstOrDefault();
+                List<PropertyInfo> keyProperties = new List<PropertyInfo>();
+                PropertyInfo[] properties = ClientTypeUtil.GetPropertiesOnType(type, false /*declaredOnly*/).ToArray();
+
+                hasProperties = properties.Length > 0;
+                typeInfo.HasProperties = hasProperties;

```

Comment:
here: typeInfo.HasProperties = hasProperties;
line 517: hasProperties = typeInfo.HasProperties.Value;

I can't understand it. Between line 462 & 517, can you let me know where do you update the "HasProperties"?

Reply:
if in 447 hasproperties is set , 517 ensures its set back to hasproperties, sicne its out param